### PR TITLE
DO NOT MERGE New way of doing frontend

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
This pull request introduces the abstract ScatteringGeneral Frontend class. This PR also provides an example implementation of the ScatteringGeneral class for the 1d case. The 1d torch and tensorflow frontends inherit from this Scattering1DGeneral class, reducing the amount of code that needs to be written up. 

If you all think that the ScatteringGeneral frontend class I've defined and the general approach of the implementation is correct, let me know as soon as possible. I will start incorporating the changes into the other PRs we have (#441 #431 #433 #434 ). 

This pull request additionally adds other changes to the code. 

One such change is changing many of the class names to their proper capitalizations (such as ScatteringTensorflow to ScatteringTensorFlow, ScatteringNumpy to Scattering NumPy). As a result, the Scattering1D class had to be adjusted for this change. I have changed the Scattering1D class to account for this. However, if users type in `Scattering1D(J=2, shape=(32,), frontend='TENSORFLOW')` or `Scattering1D(J=2, shape=(32,), frontend='Tensorflow')`, both will give you the tensorflow 1d scattering. Please let me know if y'all think this is fine. 

Another changes is some simplification of the Torch frontend code, outside the scope of the general frontend. 

The main changes are related to frontend stuff, so please focus on that. If anyone wants a UML diagram feel free to request one from me :)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -72,40 +67,81 @@ pip install kymatio
 Linux and macOS are the two officially supported operating systems.
 
 
-### GPU acceleration
+### Frontend
 
+#### NumPy
 
-To run Kymatio on a graphics processing unit (GPU), you can either use the PyTorch-style `cuda()` method to move your object to GPU. For extra speed, install the CUDA library and install the `scikit-cuda` dependency by running the following pip command:
+To explicitely call the `numpy` frontend, simply do for instance:
```
Comment (lostanlen) [inline README.md]: explicitly

File: doc/source/conf.py [hunk 2]
```diff
@@ -26,7 +26,7 @@
 # -- Project information -----------------------------------------------------
 
 project = 'kymatio'
-copyright = '2018, The Kymatio Developers'
+copyright = '2018-now, The Kymatio Developers'
```
Comment (lostanlen) [inline doc/source/conf.py]: 2018--2019

File: doc/source/index.rst [hunk 3]
```diff
@@ -33,26 +31,31 @@
...
+|pic2|  |pic1| |picDL|
 
 |pic3|  |pic4|
 
 
 Kymatio is a Python module for computing wavelet and scattering transforms.
 
-It is built on top of PyTorch, but also has a fast CUDA backend via cupy and
-skcuda.
+It is built on top of standard library in python, such as NumPy, PyTorch or TensorFlow. Our framework allows to use
```
Comment (lostanlen) [inline doc/source/index.rst]: python -> Python

File: doc/source/index.rst [hunk 4]
```diff
@@ -33,26 +31,31 @@
...
+high-level frontend to directly plug the Scattering Network to your application. It also incorporates backend which
```
Comment (lostanlen) [inline doc/source/index.rst]: scattering network

File: doc/source/index.rst [hunk 5]
```diff
@@ -33,26 +31,31 @@
...
+allows low-level tuning of the code. For instance, the PyTorch API includes a fast CUDA backend via cupy and skcuda.
 
 Use kymatio if you need a library that:
 
 * integrates wavelet scattering in a deep learning architecture,
-* supports 1-D, 2-D, and 3-D scattering transforms
+* supports 1-D, 2-D, and 3-D scattering transforms,
 * differentiable transforms for applications in generative modeling, reconstruction and more!
-* runs seamlessly on CPU and GPU hardware.
+* runs seamlessly on CPU and GPU hardware, with many deep learning API.
```
Comment (lostanlen) [inline doc/source/index.rst]: too vague

File: doc/source/userguide.rst [hunk 6]
```diff
@@ -114,15 +114,33 @@ size:
...
+Frontends
+=========
+
+You can chose a frontend via the `frontend=` arguments available in all the Scattering constructors. By default, `numpy`
+frontend is used.
+
+NumPy
+-----
+
+To call this frontend, simply do::
```
Comment (lostanlen) [inline doc/source/userguide.rst]: remove simply

File: doc/source/userguide.rst [hunk 7]
```diff
@@ -114,15 +114,33 @@ size:
...
+
+    import numpy as np
+    from kymatio import Scattering2D
+
+    scattering = Scattering2D(J=2, shape=(32, 32), frontend='numpy')
+
+This simply employs NumPy routines.
```
Comment (lostanlen) [inline doc/source/userguide.rst]: remove simply

**Comment by lostanlen** [comment]
`General` is not a good name for the abstraction.
Consider replacing by `Super` or `Interface` or `Parent` or `Base`

https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)

**Comment by edouardoyallon** [comment]
> `General` is not a good name for the abstraction.
> Consider replacing by `Super` or `Interface` or `Parent` or `Base`
> 
> https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)

I agree!

**Comment by janden** [comment]
This PR has a delta of about 5000 lines. Some of these changes have already been addressed elsewhere, and some are not relevant to the review (which is the `ScatteringGeneral` structure). Would it be possible to separate out the relevant commits here? I don't think this is reviewable as is.

**Comment by MuawizChaudhary** [comment]
> This PR has a delta of about 5000 lines. Some of these changes have already been addressed elsewhere, and some are not relevant to the review (which is the `ScatteringGeneral` structure). Would it be possible to separate out the relevant commits here? I don't think this is reviewable as is.

Yes, I separated out the `ScatteringGeneral` in #441 as `ScatteringSuper`. Later tonight I will work on incorporating the changes to the *d.

I think this pr can be closed.

**Pull request closed without merging**